bytecomp/matching.ml

```diff
@@ -246,34 +220,37 @@ end = struct
     | _ -> p
 
   let of_clause ~args cl =
-    let rec aux ((((data, view), patl), action) : General.clause) : clause =
-      match view with
-      | `Any -> (((data, `Any), patl), action)
-      | `Var (id, s) ->
-          aux (((data, `Alias (omega, id, s)), patl), action)
+    let rec aux (((p, patl), action) : General.clause) : clause =
+      let continue p (view : general_view) : clause =
+        aux (({ p with pat_desc = view }, patl), action) in
+      let stop p (view : half_simple_view) : clause =
+        (({ p with pat_desc = view }, patl), action) in
+      match p.pat_desc with
+      | `Any -> stop p `Any
+      | `Var (id, s) -> continue p (`Alias (omega, id, s))
       | `Alias (p, id, _) ->
           let arg =
             match args with
             | [] -> assert false
             | (arg, _) :: _ -> arg
           in
-          let k = Typeopt.value_kind data.env data.type_ in
+          let k = Typeopt.value_kind p.pat_env p.pat_type in
           aux ((General.view p, patl),
                bind_with_value_kind Alias (id, k) arg action)
-      | `Record ([], _) as view -> (((data, view), patl), action)
+      | `Record ([], _) as view -> stop p view
       | `Record (lbls, closed) ->
-          let full_pat = `Record (all_record_args lbls, closed) in
-          (((data, full_pat), patl), action)
-      | `Or _ as orpat -> (
-        let (data, view) =
-          General.view (simpl_orpat (General.erase (data, orpat))) in
-        match view with
-          | `Or _ as view -> (((data, view), patl), action)
-          | _ -> aux (((data, view), patl), action)
+          let full_view = `Record (all_record_args lbls, closed) in
+          stop p full_view
+      | `Or _ -> (
+        let orpat =
+          General.view (simpl_under_orpat (General.erase p)) in
+        match orpat.pat_desc with
+          | `Or _ as or_view -> continue orpat or_view
```

Comment:
That looks like an infinite loop, you should `stop`.

Reply:
Indeed. I changed it at some point and it was clearly a mistake.